firebase-firestore/ktx/src/main/kotlin/com/google/firebase/firestore/ktx/serialization/NestedMapEncoder.kt

```diff
@@ -0,0 +1,213 @@
+// limitations under the License.
+
+package com.google.firebase.firestore.ktx.serialization
+
+import kotlinx.serialization.SerializationStrategy
+import kotlinx.serialization.descriptors.SerialDescriptor
+import kotlinx.serialization.descriptors.StructureKind
+import kotlinx.serialization.descriptors.elementNames
+import kotlinx.serialization.encoding.AbstractEncoder
+import kotlinx.serialization.encoding.CompositeEncoder
+import kotlinx.serialization.modules.EmptySerializersModule
+import kotlinx.serialization.modules.SerializersModule
+import kotlinx.serialization.serializer
+
+/**
+ * The entry point of Firestore Kotlin Serialization Process. It encodes a custom @Serializable
+ * Kotlin object into a nested map of primitives. SDK users don’t need to directly interact with
+ * [NestedMapEncoder].
+ *
+ * <p>For a custom @Serializable object, at compile time, a serializer will be generated by the
+ * Kotlin serialization compiler plugin (or a custom serializer can be manually passed in). The
+ * structure information of the custom @Serializable object will be recorded inside of the
+ * serializer’s descriptor (i.e. the name/type of each property to be encoded, the annotation on
+ * each property).
+ *
+ * <p>Based on the descriptor’s information, during the encoding process, a nested map will be
+ * generated. Each property which has its own structure (i.e. a nested custom object, a nested list)
+ * will be encoded as an embedded map nested inside. At the end, the nested map at depth == 1 will
+ * be returned as the nested map representing the entire custom @Serializable object.
+ *
+ * @param map the nested map that records the encoded result of the original @Serializable object.
+ * @param depth the current encoding depth inside of the nested map.
+ * @param propertyNamesToBeEncoded a list of all property names to be encoded at the current depth.
+ */
+class NestedMapEncoder(
+    private val map: MutableMap<Int, MutableMap<String, Any?>> = mutableMapOf(),
+    private var depth: Int = 0,
+    private val propertyNamesToBeEncoded: MutableList<out Any> = mutableListOf()
+) : AbstractEncoder() {
+
+    companion object {
+        private const val ROOT_LEVEL: Int = 1
+        private const val MAX_DEPTH: Int = 500
+    }
+
+    private var elementIndex: Int = 0
+
+    init {
+        if (depth == ROOT_LEVEL) map[ROOT_LEVEL] = mutableMapOf()
+        if (depth == MAX_DEPTH) {
+            throw IllegalArgumentException(
+                "Exceeded maximum depth of $MAX_DEPTH, which likely indicates there's an object cycle"
+            )
+        }
+    }
+
+    /** Returns the final encoded nested map. */
+    fun serializedResult() = map.getValue(ROOT_LEVEL)
+
+    /** Get the field name of an enum via index, and encode it to the nested map. */
+    override fun encodeEnum(enumDescriptor: SerialDescriptor, index: Int) =
+        enumDescriptor.elementNames.toList().let { encodeValue(it[index]) }
```

Comment:
would `encodeValue(enumDescriptor.getElementName(index)` work?

Reply:
Thank you for the suggestion. It works. This is way better than my code!!!

Reply:
So `getElementName(index)` does not work?